Edge walls and End

jorisBarkema opened the pull request:
There are always walls on the edges now, and end tiles will only be placed on the edge of the maze.

There are still some issues, I'm just makign this a PR so I can get reviews.

UpdateGraphicsToMatchSurroundings is not called in the level editor wchich does start with walls now on the edges, graphical bug.

I tested the adapted mazegenerator a few times, exit seems to be working correctly but I'd like some input on this.

The Tile.Add function is kind of rough now, ideally it should show some kind of error message if you try to place a tile somewhere you can't, but I don't now how to do that.

jorisBarkema committed:
added EdgeOfTileField method

jorisBarkema committed:
reworked TileField.Add

jorisBarkema committed:
added OnCornerOfTileField helper method

jorisBarkema committed:
updated end graphics

jorisBarkema committed:
updated walltile.UpDateGraphicsToMatchSurroundings to include checking for an end tile, and UpDateGraphicsToMatchSurroundings from endtile

jorisBarkema committed:
changed mazeGenerator to only make exits on the edge of the maze

jorisBarkema committed:
mazegenerator now works properly

jorisBarkema committed:
comments

jensj12 reviewed (changes requested):
Remove all changes from the maze generator, only conflicts. Make the endTile a subclass of walltile instead of checking for endtiles in ugtms(). Make the rotation of the exit depend on the location, not surroundings.

jorisBarkema committed:
fixes

jorisBarkema commented:
Do you mean merge conflicts? Because it doesn't say there are any for me and we are going to have to change it to only make exits on the edges.

jorisBarkema committed:
fix fixes

maxmeijer42 commented:
It's not conflicted with the master but with #135

jorisBarkema commented:
aaahhh  I get it alright I'll leave it out for now, save it somewhere so we can re-add it once #135 is in master.

jorisBarkema committed:
undo mazegenerator changes

jorisBarkema committed:
tweeks

jorisBarkema committed:
tabs

jensj12 commented on MeesGame/Gameplay/GameObjects/TileField.cs:
```diff
@@ -53,6 +53,31 @@ public TileType GetType(int x, int y)
+            // If You are on the edge, but not on a corner, you can place and EndTile
+            if (OnEdgeOfTileField(x, y) && !OnCornerOfTileField(x, y) && obj.TileType == TileType.End)
+            {
+                base.Add(obj, x, y);
```
Please rework this section. Do not change the tile if the selected tile isn't allowed. Don't use more than one base.Add(). Simply return if the tile isn't allowed.

jensj12 commented on MeesGame/Gameplay/GameObjects/TileField.cs:
```diff
@@ -85,7 +110,7 @@ public Tile GetTile(int x, int y)
         }
 
-        ///this bool returns if the x and y are outside the bounds of the tilefield
+        /// This bool returns if the x and y are outside the bounds of the tilefield.
```
Summary tags are missing

jorisBarkema committed:
reworked tilefield.Add

jorisBarkema committed:
summaries added to tilefield

jensj12 committed:
Simplify TileField.Add

jorisBarkema commented:
I see the maze generator refactor is on master, is it going to stay this way for a while now? Because in that case I'll update this branch from master and implement my changes.

jensj12 commented:
If you can get your changes in before someone refactors it again (which probably won't happen again), you should implement it.

jorisBarkema committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into edge-walls

# Conflicts:
#	MeesGame/Gameplay/GameObjects/TileField.cs

jorisBarkema committed:
implement mazegeneraator changes

jorisBarkema committed:
Merge branch 'edge-walls' of https://github.com/jensj12/MeesGame.git

jorisBarkema committed:
style changes

jorisBarkema committed:
style tweaks

jorisBarkema commented:
Alright I implemented it.

jorisBarkema committed:
small bugfix

jorisBarkema commented:
There's an error when starting the game with the UGTMS in tilefield without parameter, nullreferenceException unhandled, but I have no idea where it comes from.

jensj12 commented on MeesGame/Gameplay/GameObjects/TileField.cs:
```diff
@@ -239,10 +293,10 @@ public Vector2 CellDimensions
-        /// <param name="location">The location of the Tile the player visits</param>
+        /// <summary>		
+        /// Should be called when the player enters a Tile so that the graphics can be adjusted.		
+        /// </summary>		
```
Don't add extra whitespace.

jensj12 commented on MeesGame/Gameplay/GameObjects/TilesTypes/EndTile.cs:
```diff
@@ -10,5 +10,21 @@ public override void EnterCenterOfTile(ITileFieldPlayer player)
+            if (x == 0 || x == tileField.Columns)
+            {
+                sprite = new SpriteSheet("verticalEnd");
+            }
```
Add an else to make it a horizontal exit. Don't rely on earlier states of the tile.

jensj12 commented on MeesGame/Gameplay/Generator/MazeGeneratorMain.cs:
```diff
@@ -117,8 +118,8 @@ private void MazeGenMain()
-            // Add an end tile on the best exit.
-            tiles.Add(new EndTile(), bestExit.X, bestExit.Y);
+            // Add an end tile on the edge next to the best exit.
+            tiles.Add(new EndTile(), FindEdgeTile(bestExit).X, FindEdgeTile(bestExit).Y);
```
Only call findedgetile once

jensj12 commented on MeesGame/Gameplay/Generator/MazeGeneratorInit.cs:
```diff
@@ -154,7 +154,7 @@ private void InitialiseExitSearch()
             // The exit will be placed at the hardest to reach tile, so keep an array for that.
             exitScore = new int[numRows, numCols];
-            bestExit = start;
+            bestExit = new Point(0, 1);
```
Why this change? There's an error handler in case the best exit remains the start and the start isn't on the edge. Using magic values always comes with a risk.

jensj12 commented:
Where exactly does ugtms dereference a null reference?

jorisBarkema commented:
In the foreach loop in tilefield.updategraphicstomatchsurroundings().

Only thing different from master is that I added the UGTMS override in endtile though, and when I removed that the error was still there.

jorisBarkema committed:
rebiew comments

jensj12 commented:
'Exactly' means the exact instuction and the line of code. If you know where it crashed, you which variable wasn't initialised. Then you can track down where it should have been initialised and fix the crash. In this case, tile is null, which means not every location in the tilegrid has a tile. The callstack tells us that the initialisation failed in the constructor of editorlevel (as it called ugtms from there). It tries to place floortiles on the edge, which isn't allowed.

jorisBarkema committed:
fixed tilefield loading error

jorisBarkema commented:
error has been fixed

jensj12 commented:
OnEdgeOfTileField doesn't work correctly on the right and lower side of the maze. Visible when starting the editor. Also, those three functions contain more code than necessary. Checking if a bool is true and then returning true (and false otherwise) isn't needed.

jorisBarkema committed:
minor bugfix in tilefield

jorisBarkema commented:
Oops, silly mistake, fixed it.

LisanneW commented:
Starting the level editor doesn't immediatly call ugtms() on all tiles. The orientation of the exit is incorrect on the right side of the maze.

jorisBarkema committed:
editor level ugtms fix and small endtile.ugtms fix

jorisBarkema commented:
should be good now

LisanneW committed:
Correct solution to ugtms
